src/crystal/main.cr

```diff
@@ -0,0 +1,112 @@
+lib LibCrystalMain
+  @[Raises]
+  fun __crystal_main(argc : Int32, argv : UInt8**)
+end
+
+# :nodoc:
+def _crystal_main(argc : Int32, argv : UInt8**)
+  # TODO: remove this method and embed this inside
+  # Crystal.main. A bug in Crystal 0.23.1 prevents invoking
+  # __crystal_main from anywhere except the top level.
+  LibCrystalMain.__crystal_main(argc, argv)
+end
+
+module Crystal
+  # Defines the main routine run by normal Crystal programs:
+  #
+  # - Initializes the GC
+  # - Invokes the given *block*
+  # - Handles unhandled exceptions
+  # - Invokes `at_exit` handlers
+  # - Flushes `STDOUT` and `STDERR`
+  #
+  # This method can be invoked if you need to define a custom
+  # main (as in C main) function, doing all the above steps.
+  #
+  # For example:
+  #
+  # ```
+  # fun main(argc : Int32, argv : UInt8**) : Int32
+  #   Crystal.main do
+  #     time = Time.now
+  #     Crystal.main_user_code(argc, argv)
+  #     puts "Time to execute program: #{Time.now - time}"
+  #   end
+  # end
+  # ```
+  #
+  # Note that the above is really just an example, almost the
+  # same can be accomplished with `at_exit`. But in some cases
+  # redefinition of C's main is needed.
+  def self.main(&block)
+    GC.init
+
+    status =
+      begin
+        yield
+        0
+      rescue ex
+        1
+      end
+
+    AtExitHandlers.run status
+    ex.inspect_with_backtrace STDERR if ex
+    STDOUT.flush
+    STDERR.flush
+    status
+  end
+
+  # Main method run by all Crystal programs at startup.
+  #
+  # This setups up the GC, invokes your program, rescuing
+  # any handled exception, and then runs `at_exit` handlers.
+  #
+  # This method is automatically invoked for you, so you
+  # don't need to invoke it.
+  #
+  # However, if you need to define a special main C function,
+  # you can redefine main and invoke `Crystal.main` from it:
+  #
+  # ```
+  # fun main(argc : Int32, argv : UInt8**) : Int32
+  #   # some setup before Crystal main
+  #   Crystal.main(argc, argv)
+  #   # some cleanup logic after Crystal main
+  # end
+  # ```
+  #
+  # The `Crystal.main` can also be passed as a callback:
+  #
+  # ```
+  # fun main(argc : Int32, argv : UInt8**) : Int32
+  #   LibFoo.init_foo_and_invoke_main(argc, argv, ->Crystal.main)
+  # end
+  # ```
+  #
+  # Note that before `Crystal.main` is invoked the GC
+  # is not setup yet, so nothing that allocates memory
+  # in Crystal (like `new` for classes) can be used.
+  def self.main(argc : Int32, argv : UInt8**)
+    main do
+      main_user_code(argc, argv)
+    end
+  end
+
+  # Executes the main user code. This normally is executed
+  # after initializing the GC and before executing `at_exit` handlers.
+  #
+  # You should never invoke this method unless you need to
```

Comment:
Would it be possible to raise early if this is executed twice? Seems like a bit of an edge case but might be worth protecting. This could be another PR though.